lib/lz4frame.c

```diff
@@ -904,7 +906,7 @@ size_t LZ4F_compressUpdate(LZ4F_cctx* cctxPtr,
     }
 
     /* keep tmpIn within limits */
-    if ((cctxPtr->tmpIn + blockSize) > (cctxPtr->tmpBuff + cctxPtr->maxBufferSize)   /* necessarily LZ4F_blockLinked && lastBlockCompressed==fromTmpBuffer */
+    if (((uptrval) cctxPtr->tmpIn + blockSize) > ((uptrval) cctxPtr->tmpBuff + cctxPtr->maxBufferSize)   /* necessarily LZ4F_blockLinked && lastBlockCompressed==fromTmpBuffer */
```

Comment:
`make usan` reports

for `cctxPtr->tmpIn == NULL`

```
../lib/lz4frame.c:907:25: runtime error: applying non-zero offset 65536 to null pointer
../lib/lz4frame.c:907:25: runtime error: applying non-zero offset 262144 to null pointer
../lib/lz4frame.c:907:25: runtime error: applying non-zero offset 1048576 to null pointer
../lib/lz4frame.c:907:25: runtime error: applying non-zero offset 4194304 to null pointer
```

for `cctxPtr->tmpBuff`

```
../lib/lz4frame.c:907:58: runtime error: applying zero offset to null pointer
../lib/lz4frame.c:907:58: runtime error: applying non-zero offset 5242880 to null pointer
```

Reply:
investigating

Reply:
OK, I've got one scenario.

When using `lz4frame` in streaming mode,
with `autoflush` enabled and independent blocks mode,
then there is no need for any internal buffer, 
since input is always entirely consumed and flushed, 
and no history persists between invocations.

In which case `cctx->tmpBuffer` can be `NULL`.

After that, applying pointer arithmetic on this buffer is UB.
Of course, it's also useless, because it means that there is nothing to store.

I presume it will be enough to swap the `&&` test order.

Reply:
There is a fix for this issue at :  https://github.com/lz4/lz4/pull/986

I don't see any remaining UB coming from this line after this patch.

Reply:
I've also confirmed.  This UB is eliminated by #986.

```
$ git log -1
commit 8e46846287c60bb34ab8157113715b94e39261b9 (HEAD -> fixub907, upstream/fixub907)
Author: Yann Collet <cyan@fb.com>
Date:   Thu May 27 22:59:22 2021 -0700

    fix UB of lz4frame:907

    now line 912
    by ensuring pointer arithmetic is only performed
    if there is a reason for an internal buffer to be used.
$ make usan MOREFLAGS='-Wcomma -Werror' 2> make-usan-fixub907
$ grep "lz4frame.c" make-usan-fixub907
  646   ../lib/lz4frame.c:1564:24: runtime error: applying zero offset to null pointer
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior ../lib/lz4frame.c:1564:24 in
```